### PR TITLE
Project Builder: Remove CoffeeScript from Workflow Editor

### PR DESCRIPTION
**Pull request by shaunanoordin**
## PR Overview

Affects: Project Builder
🆕 ⭐ **Context:** as part of the migration to FEM, we plan to build a "Steps/Pages Editor" in the (PFE) Project Builder. In preparation for that, the old Workflow Editor page needs to be converted into modern code.

This PR converts the Workflow Editor (both **PFE Lab** and **FEM Lab** versions) from CoffeeScript to JavaScript.

Feature/Functional Changes:
- Nothing! Workflow Editor _should_ function as it always had.

Code Changes:
- First pass: `workflow.cjsx` converted to `workflow.jsx` via [decaffeinate](https://decaffeinate-project.org/repl/)
  - Options used in the auto-conversion process: `--use-cs2` (necessary for our code) and `--optional-chaining` (because we can use`?.` in our JS)
- `require` converted to `import`
- `EditWorkflowPageWrapper` converted to a functional React component
- `EditWorkflowPage` converted to a React component
  - _This thing is so huge, that converting to a functional component would best be done at a separate PR_ 
- Syncing PFE Lab -> FEM Lab changes from #6212
  - 6212 introduces a new function in the "workflow actions" for copying WFs, but only applied it to PFE Lab's WF Editor.

Notes:
- Changes _should_ be mirrored across the PFE Lab's version of Workflow Editor, and FEM Lab's version.
- Event handlers had to be updated with `onEvent={this.handleEvent.bind(this)}`, since previously the `@` in `onEvent={@handleEvent}` not only substituted for 'this', but also seems to have had an implicit .bind()

### Status

WIP. Please stay tuned for testing notes, because this PR will benefit from multiple eyes reviewing/testing the Workflow Editor to ensure functional parity.

### CODE REVIEW

**Comment by eatyourgreens** [comment]
`.bind()` creates a new function each time it runs, so avoid running it when a component renders.

**Comment by shaunanoordin** [comment]
> .bind() creates a new function each time it runs, so avoid running it when a component renders.

Noted, thanks. This should be sorted when we convert EditWorkflowPage into a functional component.

**Comment by eatyourgreens** [comment]
For now, binding your event handlers in the constructor should sort it. 

```js
this.handleEvent = this.handleEvent.bind(this);
```

The handler is bound once, when the class is instantiated, and can be safely referenced as `this.handleEvent` without changing the `this` context.

**Comment by shaunanoordin** [comment]
OK, I think this is ready for testing.

### Testing

Testing Notes:
- The idea here is that the (new, non-CoffeeScript) Workflow Editor should behave _the same_ as the (old, CoffeeScript) Workflow Editor. We're checking for parity.
- ❗ We need to test twice, once for the 🍎 **FEM Lab** version and once for the 🍊 **PFE Lab** version.
- Recommended testing plan (below) has an expected testing time of 5-10 mins total.

Test URLs:
- 🍎 FEM Lab: https://pr-6220.pfe-preview.zooniverse.org/lab/1966/workflows/3697?env=staging&femLab=true
- 🍊 PFE Lab: https://pr-6220.pfe-preview.zooniverse.org/lab/1966/workflows/3697?env=staging&pfeLab=true
- You can use your own project/WF, but please be sure to use 🍊 `?pfeLab=true` or 🍎 `?femLab=true` to test for PFE Lab and FEM Lab specifically.
- You can tell you're on 🍎 FEM Lab when the "Set Annotation Persistence", "Set As Default", etc sections are disabled and marked as "deprecated"

Testing Steps:
- **Basic:**
  - Open the link. You should be able to view the Workflow Editor page and nothing should crash.
  - Change the _Workflow name._ This should work as expected. (TSWAE)
  - Select and/or un-select any number of SbjSets as _Associated Subject Sets._ TSWAE.
- **Adding Tasks:**
  - Click "_Add a Task"._ You should see multiple options.
  - Choose a Question/Text/Drawing/etc Task. The new Task should be added to the WF.
  - Optional: edit the Task.
  - Optional: add more Tasks.
  - Optional: change First Task.
- **Deleting Tasks:** select a Task, then click "Delete this Task". TSWAE.
- **Toggle Various Flags:** change the values of the following options/settings:
  - _Persist Annotations_ (PFE Lab only)
  - _Default Workflow_ (PFE Lab only)
  - _Associated Tutorial_
  - _Hide classification summaries_ (PFE Lab only)
  - _Pan and Zoom_ (PFE Lab only)
  - _Multi Image Options_ (PFE Lab only)
  - _Allow users to flip image colour_ 
  - _Subject Retirement_
- ❗ **Confirmation:**
  - After you've made all the above changes, either RELOAD/REFRESH the page, or go back to the Workflows page, then return to the workflow in question.
  - Look through your workflow again. You're checking to make sure that all the changes you've made to the workflow are actually registered in the database.
- **Copy Workflow:**
  - click the "Copy icon" (next to the top-most Workflow name) to open the "Copy Workflow" menu.
  - Enter a new WF name and click OK. You should now be editing a copy of your previous WF.

Additional Tests:
- Create a new workflow from the Workflows Page. The newly created workflow should be immediately displayed in the Workflow Editor

### Status

This PR is now ready for review, and in its current stage, should serve as a good enough baseline for future modifications/maintenance. (I'm looking at you, Pages/Steps Editor.)

@lcjohnso and/or @snblickhan , if you can spare 5-10 mins to help test this from a project owner's PoV, it'd be much appreciated. I want to make sure I didn't miss anything obvious from a POwner's POV.

I plan to follow this up with another "optimisation PR", so please let me know of any major suggestions. So far, on my list:
- convert EditWorkflowPage to a functional component.
- Reduce use of eventHandler.bind() at the render stage, _where possible._ (I'll be honest, there are some pretty complicated uses of .bind() here that I really prefer not to touch.)
- Remove PromiseRenderer.
- (super ambitious) break EditWorkflowPage into smaller components.

File: app/pages/lab/workflow.jsx
Comment (eatyourgreens) [inline app/pages/lab/workflow.jsx:148]: Why is this declared as an anonymous function that runs automatically? It’s very hard to read.

File: app/pages/lab/workflow.jsx
Comment (eatyourgreens) [inline app/pages/lab/workflow.jsx:365]: Rendering the return value of a function like this means that `renderTutorials` can probably be extracted into a component.

File: app/pages/lab/workflow.jsx
Comment (eatyourgreens) [inline app/pages/lab/workflow.jsx:1035]: This `PromiseRenderer` should be replaced by a `useState` hook that stores the workflow in component state after fetching it from the API. Otherwise, you’re making network requests here even after you’ve already fetched the workflow. 

That same comment holds for the promises that fetch the tutorials and mini courses.

File: app/pages/lab/workflow.jsx
Comment (eatyourgreens) [inline app/pages/lab/workflow.jsx:163]: Another auto-running anonymous function, nested inside the first auto running anonymous function. I think this is going to be really hard to read and debug.

Comment (eatyourgreens) [reply app/pages/lab/workflow.jsx:163]: I think this function could be extracted from the component and declared at the top of the file, which would make the code much easier to understand.

File: app/pages/lab/workflow.jsx
Comment (eatyourgreens) [inline app/pages/lab/workflow.jsx:1]: The `React` global isn’t needed. 
```suggestion
```

Comment (eatyourgreens) [reply app/pages/lab/workflow.jsx:1]: Probably better to just import any specific React exports that you need eg.
```js
import { Component } from ‘react’;
```

**Comment by eatyourgreens** [comment]
ES imports must use the full file path, including the extension. See the imports in #6219. It might be worthwhile doing that here too.

File: app/pages/lab/workflow.jsx
Comment (eatyourgreens) [reply app/pages/lab/workflow.jsx:1035]: I’ve included an example of this in #6219, when the project editor loads the project.

**Comment by eatyourgreens** [comment]
This looks like a good start at removing CoffeeScript but I don’t think the JS code is quite ready to use in production. 

I’d like to see:
- removal of the auto-running anonymous closures, particularly the nested anonymous functions. I think those will be very hard to maintain and debug. 
- remove `PromiseRenderer`, as recommended [in the repo Readme](https://github.com/zooniverse/Panoptes-Front-End/blob/master/README.md#async-helper-components). The problem with this component is that it sends an API request on every render. `React.Suspense` might be a good replacement, which displays a placeholder until a promise has resolved.

**Comment by shaunanoordin** [comment]
Thanks for the review, Jim.

- As mentioned in the [update](https://github.com/zooniverse/Panoptes-Front-End/pull/6220#issuecomment-1284521205), (under "Status"), I'm not planning on changing Promise Renderer in this PR. It _is_ planned for the next PR, which is on this/next week's ToDo.
  - Removing the anonymous ad-hoc functions in the middle of the code also falls into this "next PR" category.
- Noted on removing `React` global. 👌 Update incoming.
- Noted on using full paths for `import`s 👌 Update incoming.
  - e.g. `import AutoSave from '../../components/auto-save';` -> `import AutoSave from '../../components/auto-save.coffee';`

Thank you for helping note the many things that need to be improved, but please note that I want to approach this CoffeeScript conversion in discrete steps instead one big chunk.

The goal of this PR is to do the minimum possible move away from CoffeeScript while maintaining functional parity, while a second (or third) optimisation PR(s) would come in to sweep up those problems you noted (and more - I just noticed those Array.from()s). My rationale is that with more PRs, it'll be easier to figure out what I'm optimising (and spot any errors, esp. across two Labs), and to roll back to a safer position if my update went wrong.

File: app/pages/lab/workflow.jsx
Comment (eatyourgreens) [inline app/pages/lab/workflow.jsx:109]: `this.showCreateWorkflow` should be declared in the constructor, following the pattern here:
https://github.com/zooniverse/Panoptes-Front-End/blob/6b128109d64c4d30429e37f44dc62e3f6fa3ed64/app/classifier/tasks/highlighter/index.jsx#L47-L52

Comment (eatyourgreens) [reply app/pages/lab/workflow.jsx:109]: CoffeeScript did this for us, with `@`, but that’s been removed in this PR. Now, new event handlers are generated on every render. Each time you toggle a checkbox, or type in a text box, a new set of event handlers is added to the component. That leads to memory leaks in the browser.

Comment (eatyourgreens) [reply app/pages/lab/workflow.jsx:109]: I can go through the code and make this change as review suggestions, or open a PR. Either way, it’s not a lot of work.

Comment (eatyourgreens) [reply app/pages/lab/workflow.jsx:109]: Since this changes the behaviour of the component from the CoffeeScript version, I’d prefer this change to be made in this PR, not after.

Comment (eatyourgreens) [reply app/pages/lab/workflow.jsx:109]: #6225.

File: app/pages/lab/workflow.jsx
Comment (eatyourgreens) [reply app/pages/lab/workflow.jsx:1035]: #6226 fetches the workflow with a hook, which should only run when the workflow ID changes.

**Comment by eatyourgreens** [comment]
I'd at least like to see #6225 and #6226 merged into this PR before these changes are put in front of real project teams. They're both pretty small PRs, but should improve performance.

**Comment by eatyourgreens** [comment]
I should also note that `createReactClass` automatically bound the `this` context for a React component, when the component mounts. Now that `createReactClass` has been removed, #6225 will restore parity with the original workflow editor.

**Comment by eatyourgreens** [comment]
#6226 subscribes the page to workflow changes, which fire when `workflow.update()` runs.

**Comment by eatyourgreens** [comment]
I've started on #6228 to clean up the decaffeinate code for tutorials and mini-courses. Something I noticed, working on that, is that PFE's linting rules haven't been applied to this branch. The decaffeinate output might benefit from being run through PFE's default ESLint rules, maybe in a separate PR?

**Comment by eatyourgreens** [comment]
Sure. #6228 isn't needed for this PR. It's the first stage in manually cleaning up the decaffeinate conversion.

I would hold off on deploying the changes here from staging to production, at least until the JS has been cleaned up. The code is still pretty hard to follow.

**Comment by shaunanoordin** [comment]
What's the criteria for the JS to be "cleaned up" enough for this PR to be ready to deploy/ready to merge? The Workflow Editor, in its original CoffeeScript form, is equally hard to follow. (Unless you're telling me this decaffeinate attempt is actually making the code readability _worse,_ in which, this PR definitely can't be merged.)

My concern is that if we wait to keep polishing up the Workflow Editor to more modern coding standards, we'll have a large monolithic PR that sits around for weeks, which can be problematic if/when we make bug fixes/changes/etc in master that then need to be synced to the post-CoffeeScript editor. I'd very much prefer to commit safe-but-unoptimised changes we can iterate on.

**Comment by eatyourgreens** [comment]
I’m happy with this PR right now, at least enough to merge and deploy to staging. #6228’s base branch should automatically change once this merges. If not, the rebase should be trivial. 

I am arguing that the JS version is worse than the CoffeeScript version, but we can continue to iterate and clean it on PFE staging. I’d rather do the manual cleanup in small, manageable PRs than have this one continue to grow. #6228 is the first of those small PRs.

**Comment by eatyourgreens** [comment]
I’ve updated #6228 to make it explicitly clear that this should merge first.

**Comment by eatyourgreens** [comment]
Merging this PR will also get us one step closer to closing #3216, which would be a big step forward.

File: app/pages/lab/workflow.jsx
Comment (eatyourgreens) [reply app/pages/lab/workflow.jsx]: Fixed now, so we have parity with the old `createReactClass` code.

File: app/pages/lab/workflow.jsx
Comment (eatyourgreens) [reply app/pages/lab/workflow.jsx:385]: Moved into #6228.

File: app/pages/lab/workflow.jsx
Comment (eatyourgreens) [reply app/pages/lab/workflow.jsx:168]: This is decaffeinate’s answer to mapping arrays.

File: app/pages/lab-fem/workflow.jsx
Comment (eatyourgreens) [inline app/pages/lab-fem/workflow.jsx:203]: This line rebinds `this.setState`.

Comment (eatyourgreens) [reply app/pages/lab-fem/workflow.jsx:203]: The original did this too. 😳

https://github.com/zooniverse/Panoptes-Front-End/blob/80441b7e1a328eb3ed32bc8374ea0b7033a74b55/app/pages/lab/workflow.cjsx#L151

**Comment by shaunanoordin** [comment]
## PR Update

Testing complete. This PR is now ready for review.

### Testing

"Edit Workflow" Testing Steps:
- Select a workflow in either PFE Lab mode. (After all the steps are completed, repeat with FEM Lab mode.)
  - Example URL 1: https://local.zooniverse.org:3735/lab/1986/workflows/3714?pfeLab=true
  - Example URL 2: https://local.zooniverse.org:3735/lab/1986/workflows/3714?femLab=true
- Change the following then reload the page to confirm it sticks:
  - WF Title, Associated Subject Sets, Tutorial
  - allow users to flip image color, subject retirement
  - (PFE-Lab only:) Persist annotations, default workflow (note: option only enabled if project has 2+ workflows), hide classification summaries, pan and zoom, multi-image options
- Misc:
  - Check that the "Contact Us" link works
  - Check that we can change the First Task (if we have 1+ Tasks)
  - Check that we can "Test This Workflow" and "View Project"
- **Question Task,** check that the following all work and changes stick after page reload:
  - Add a Question Task, edit main text, and edit help text
  - Toggle allow hiding marks/allow multiple/required
  - Add Answers
  - Set "Next Task" for each Answer (only when 'allow multiple' disabled)
  - Set "Next Task" for Task as a whole (only when 'allow multiple' enabled)
  - Delete the Task
  - ⚠️ Possible Quirk?: when creating a second Question Task, the Allow Hiding Marks value seems to match the first? I need to take a look at this later.
- **Drawing Task:**
  - Add a Drawing Task, edit main text, and edit help text
  - Toggle allow hiding marks
  - Add a Drawing tool, change type/color/min/max/size (where applicable)
    - Add some subtasks: question/slider/text
    - Delete some subtasks
    - ⚠️ Possible quirk: the "Add new sub-task" button can be difficult to interact with, as it may render with its bottom half "clipped off" by the bounding modal box.
  - Add more Drawing Tools, then delete some Drawing Tools
  - Change "Next Task"
  - Delete the Task
- **Text Task:**
  - Add a Text Task, edit main text, and edit help text
  - Toggle 'required'
  - Toggle metadata tags
  - Change "Next Task"
  - Delete the Task
- **Survey Task:**
  - Add a Survey Task
  - Using provided sample data, add choices, characteristics, confused pairs, and questions
  - Check that the data appears under "Applied Survey Data"
  - Toggle 'require at least one identification' and 'always show thumbnails'
  - Change "Next Task"
  - Delete the Task
- **Transcription Task:**
  - Add a Transcription Task (comes in a pair of tasks)
  - For Task 1 (drawing),
    - edit main text, edit help text, and edit sub-task text
    - toggle metadata tags
  - For Task 2 (question),
    - edit main text and edit help text
    - toggle allow hiding marks/allow multiplpe/required
  - Delete either Task 1 or Task 2. The other Task should also be deleted.

EXTENSIVE Sanity Checks:

- I can go to the main [Project Builder Page](https://local.zooniverse.org:3735/lab) and create a new project
- I can edit various input fields in Project Details
  - This includes name, description, introduction, and "Volunteers can choose which workflow they work on".
  - I can reload the page the confirm changes.
  - ⚠️ ❔ There may be some **quirks or bugs** with uploading Avatar and Background images. Sometimes, after selecting & uploading a file, the new image either isn't saved to the DB or isn't immediately reflected on the page. Further investigation required, may open separate issue.
- I can edit the following sub-sections:
  - About: can add & update text for each page
  - Collaborators: can add, edit, remove users
  - Field Guide: can create 1-page FG
  - Tutorial: can create 1-step Tutorial
  - Media: can upload 2 images
  - Visibility: can make public & live
  - Workflows: can make a new WF, set WFs as active/inactive, and change each WF's statistics setting
  - Subject Sets: can make a new SbjSet and upload 10 images

### Status

Ready for review

**Comment by eatyourgreens** [comment]
I haven’t seen any problems on this branch while editing my test project, so I’m happy to merge this. 

However, I’m not a big user of the project builder. @shaunanoordin do you want a review from someone who is more familiar with the project builder, before this goes out to staging? I don’t want to merge prematurely.

**Review by eatyourgreens — APPROVED** [review]

**Comment by shaunanoordin** [comment]
Thanks for the approve, Jim. 👍 

> do you want a review from someone who is more familiar with the project builder, before this goes out to staging? I don’t want to merge prematurely.

Absolutely, I think that's a great idea. I'll ping Sam & Cliff on Slack to see if they can get eyes on this in the next few days.

(Note to whoever's reading this: the test URL is https://pr-6220.pfe-preview.zooniverse.org/lab )

**Comment by trouille** [comment]
Just noting that I'll have some time tomorrow (Wednesday 3:15-4pm CT) to review. I'll post here feedback/comments.

**Review by lcjohnso — APPROVED** [review]
I tested basic workflow and task editing and encountered no issues related to this PR.

I did identify an issue related to caching and API responses / page state, but I can replicate the issue on the deployed prod version of the project builder, so not related.  That issue is documented here: https://github.com/zooniverse/Panoptes-Front-End/issues/6230

**Comment by trouille** [comment]
Had a chance to review this afternoon as well. From what I tested, it's looking good; working as expected. 

I tested creating a new project with two new workflows. Tested having volunteers choose which workflows or not. One workflow included two question tasks. The other was a marking task with a subtask. Tested creating a tutorial and linking it to a workflow. All working as expected. 

Shaun - you noted above, "Possible Quirk?: when creating a second Question Task, the Allow Hiding Marks value seems to match the first? I need to take a look at this later.". Playing around w/ the Question Tasks, seems fine that 'Allow Hiding Marks' is an option, but is it used / relevant (in a question task, what marks would you be removing?). 

You also noted, Possible quirk: the "Add new sub-task" button can be difficult to interact with, as it may render with its bottom half "clipped off" by the bounding modal box. I'll just note that I didn't come across this issue. The subtask rendered ask expected and would overflow outside the subject viewer as expected. 

Final note here - you also noted, "quirks or bugs with uploading Avatar and Background images. Sometimes, after selecting & uploading a file, the new image either isn't saved to the DB or isn't immediately reflected on the page." Yes, agree with this. Continue to see a delay in when the avatar and background images appear.

**Comment by shaunanoordin** [comment]
Thanks, Cliff and Laura! 👍 

Cliff: noted on the caching issue with 6230, I'll look into replicating that on my local.

Laura: thanks also for looking into those quirks. (1) "is Allow Hiding Marks" relevant, e.g. on question task?" 🤔 good point, let me dig into that further. (2) "(there is still a) delay in when the avatar and background images." this is good confirmation - I'll open up an issue for this shortly.

Alright, let's get this merged and continue with the cleanup! I really appreciate everyone's help reviewing & updating this mega refactor. 👌

**Pull request merged**
